### PR TITLE
reorganize a bit of early library and init code

### PR DESCRIPTION
**Pull request by JeffBezanson**
base.jl had become kind of a random mix of things. This moves the exception types to related files, and the other bits to their own files. The exception types referenced by the runtime also arguably should be in boot.jl, so they're moved there. As a side effect we can remove `jl_get_system_hooks`.

(Also I believe there might have been cases where `jl_complex_type` was not getting set early enough, since e.g. `Complex` was not defined when Base called `jl_get_system_hooks`. In case that was a problem, it should be fixed by this PR too.)

### CODE REVIEW

**Comment by tkelman** [comment]
We're trying to eventually delete helpdb, not add things back there...

**Comment by JeffBezanson** [comment]
I put them in helpdb because the types are defined in boot.jl, where the doc system doesn't work. It's quite likely that after everything else is moved out of helpdb, we'll be left with a file just to document the things in boot.jl.

**Comment by tkelman** [comment]
we already have that, it's called basedocs.jl

**Comment by JeffBezanson** [comment]
Ok, should everything defined in boot.jl be moved to basedocs.jl?

File: base/docs/basedocs.jl [hunk 1]
```diff
@@ -727,4 +729,361 @@ true
...
+    OverflowError(msg)
+
+The result of an expression is too large for the specified type and will cause a wraparound.
+"""
+OverflowError
+
+"""
+    TypeError(func::Symbol, context::AbstractString, expected::Type, got)
+
+A type assertion failure, or calling an intrinsic function with an incorrect argument type.
```
Comment (tkelman) [inline base/docs/basedocs.jl]: hm, that's not a sentence (can be fixed later)

File: base/boot.jl [hunk 2]
```diff
@@ -244,6 +244,40 @@ struct OverflowError <: Exception
...
+    MethodError(@nospecialize(f), @nospecialize(args), world::UInt) = new(f, args, world)
+end
+const typemax_UInt = ccall(:jl_typemax_uint, Any, (Any,), UInt)
+MethodError(@nospecialize(f), @nospecialize(args)) = MethodError(f, args, typemax_UInt)
+
+mutable struct AssertionError <: Exception
+    msg::AbstractString
+    AssertionError() = new("")
+    AssertionError(msg) = new(msg)
+end
```
Comment (musm) [inline base/boot.jl]: Unrelated but shouldn't these be  ?
```
mutable struct AssertionError <: Exception
     msg::String
     AssertionError() = new("")
     AssertionError(msg::AbstractString) = new(msg)
 end
```

**Pull request merged**